### PR TITLE
CORE-15909: make group parameters accessible in flow & contract verification

### PR DESCRIPTION
**Pull request by jennyang-r3**
### Overview
This PR
- renamed `CurrentGroupParametersService` ->  `GroupParametersLookup`
- separation of interface for internal use and public API
- changed tests accordingly
- group param is accessible from flow using `GroupParametersLookup`
- group param is accessible via transaction in contract verification
- class evolution and handle preventing from accessing new field from 5.0

**Local test result is successful**
- group parameters are accessible in verification: https://gradle.dev.r3.com/s/ddopbxgvg5xmw
- group parameters service is injected to a flow successfully: https://gradle.dev.r3.com/s/vgv2qn7oavwlo

**E2E tests are written** in [this PR 
](https://github.com/corda/corda-e2e-tests/pull/205)

**Public API implementation:** https://github.com/corda/corda-api/pull/1202

### CODE REVIEW

File: libs/ledger/ledger-utxo-data/build.gradle [hunk 1]
```diff
@@ -14,6 +14,7 @@ dependencies {
 
     implementation project(':libs:crypto:crypto-core')
     implementation project(':libs:ledger:ledger-common-data')
+    implementation project(':libs:membership:membership-common')
```
Comment (vlajos) [inline libs/ledger/ledger-utxo-data/build.gradle:17]: Why is this needed?

Comment (jennyang-r3) [reply libs/ledger/ledger-utxo-data/build.gradle:17]: ignore this, I was thinking of bringing the group parameters to `utxoLedgerTransaction` but haven't done it yet so

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/groupparameters/GroupParametersServiceImpl.kt [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        val groupReader = membershipGroupReaderProvider.getGroupReader(holdingIdentity)
+        val groupParameters = requireNotNull(groupReader.groupParameters) {
+            "groupParameters could not be accessed."
+        }
+        return groupParameters
+    }
+
+    private val holdingIdentity: HoldingIdentity
+        get() =
+            currentSandboxGroupContext.get().virtualNodeContext.holdingIdentity
```
Comment (chrisr3) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/groupparameters/GroupParametersServiceImpl.kt]: Is this _really_ the only way for a `PROTOTYPE` sandbox service to obtain its own `HoldingIdentity`? :thinking:

Comment (vlajos) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/groupparameters/GroupParametersServiceImpl.kt]: This pattern looks like repeated quite a few times in the code.

File: libs/ledger/ledger-utxo-data/build.gradle [hunk 3]
```diff
@@ -15,7 +15,10 @@ dependencies {
     implementation project(':libs:crypto:crypto-core')
     implementation project(':libs:ledger:ledger-common-data')
     implementation project(':libs:membership:membership-common')
+    implementation project(':libs:membership:membership-impl')
     implementation project(':libs:sandbox-types')
+    implementation project(':libs:serialization:serialization-amqp')
```
Comment (chrisr3) [inline libs/ledger/ledger-utxo-data/build.gradle]: Why do you need `membership-impl` or `serialization-amqp`? You shouldn't need anything other than what you need to compile this module.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/verifier/UtxoLedgerTransactionVerificationServiceImpl.kt [hunk 4]
```diff
@@ -79,7 +79,7 @@ class UtxoLedgerTransactionVerificationServiceImpl @Activate constructor(
     private fun fetchAndVerifySignedGroupParameters(transaction: UtxoLedgerTransaction): SignedGroupParameters {
```
Comment (vlajos) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/verifier/UtxoLedgerTransactionVerificationServiceImpl.kt]: The `if (currentGroupParameters.hash.toString() == membershipGroupParametersHashString) {` shortcut could be moved into UtxoLedgerTransactionFactoryImpl's `getGroupParameters()`.
Then this function could only verify since the input transaction's groupParameters would be ready.
Also, I think currentGroupParametersService would not be needed anymore in this class.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/verifier/UtxoLedgerTransactionVerificationServiceImpl.kt [hunk 5]
```diff
@@ -99,7 +99,21 @@ class UtxoLedgerTransactionVerificationServiceImpl @Activate constructor(
...
-            UtxoLedgerTransactionContainer(wireTransaction, inputStateAndRefs, referenceStateAndRefs)
+            val membershipGroupParametersHashString = this.getMembershipGroupParametersHash()
+            val groupParameters = requireNotNull(
+                utxoLedgerGroupParametersPersistenceService.find(
+                    parseSecureHash(membershipGroupParametersHashString)
+                )
+            ) {
+                "Signed group parameters $membershipGroupParametersHashString related to the transaction" +
+                        " ${wireTransaction.id} cannot be accessed."
+            }
```
Comment (vlajos) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/verifier/UtxoLedgerTransactionVerificationServiceImpl.kt]: AFAIK the groupParameters are already available in the transaction's related property.

File: components/ledger/ledger-verification/build.gradle [hunk 6]
```diff
@@ -45,6 +45,7 @@ dependencies {
     implementation project(':libs:tracing')
     implementation project(':libs:utilities')
     implementation project(':libs:virtual-node:sandbox-group-context')
+    implementation project(':libs:membership:membership-common')
```
Comment (vlajos) [inline components/ledger/ledger-verification/build.gradle:48]: Why is this needed?

File: libs/ledger/ledger-utxo-data/build.gradle [hunk 7]
```diff
@@ -14,7 +14,10 @@ dependencies {
 
     implementation project(':libs:crypto:crypto-core')
     implementation project(':libs:ledger:ledger-common-data')
+    implementation project(':libs:membership:membership-common')
+    implementation project(':libs:membership:membership-impl')
     implementation project(':libs:sandbox-types')
+    implementation project(':libs:serialization:serialization-checkpoint-api')
```
Comment (vlajos) [inline libs/ledger/ledger-utxo-data/build.gradle]: AFAIK this is needed only for Kryo serializer. And those should not be in ledger-utxo-data.
Is this used?

File: components/virtual-node/sandbox-group-context-service/test.bndrun [hunk 8]
```diff
@@ -8,7 +8,8 @@
     osgi.service;objectClass:List<String>='net.corda.configuration.read.ConfigurationReadService';effective=active,\
     osgi.service;objectClass:List<String>='net.corda.lifecycle.LifecycleCoordinatorFactory';effective:=active,\
     osgi.service;objectClass:List<String>='net.corda.messaging.api.subscription.factory.SubscriptionFactory';effective:=active,\
-    osgi.service;objectClass:List<String>='org.osgi.service.condpermadmin.ConditionalPermissionAdmin';effective:=active
+    osgi.service;objectClass:List<String>='org.osgi.service.condpermadmin.ConditionalPermissionAdmin';effective:=active,\
+    osgi.service;objectClass:List<String>='net.corda.avro.serialization.CordaAvroSerializationFactory';effective:=active
```
Comment (vlajos) [inline components/virtual-node/sandbox-group-context-service/test.bndrun:12]: What do these do?

File: simulator/runtime/src/main/kotlin/net/corda/simulator/runtime/ledger/utxo/UtxoLedgerTransactionBase.kt [hunk 9]
```diff
@@ -151,6 +152,9 @@ data class UtxoLedgerTransactionBase(
         return emptyList() // TODO Not yet Implemented
     }
 
+    override fun getGroupParameters(): GroupParameters {
+        TODO("Not implemented")
```
Comment (vlajos) [inline simulator/runtime/src/main/kotlin/net/corda/simulator/runtime/ledger/utxo/UtxoLedgerTransactionBase.kt:158]: This may need to be checked with the DevEx team.

File: libs/ledger/ledger-utxo-data/src/main/kotlin/net/corda/ledger/utxo/data/transaction/UtxoLedgerTransactionContainer.kt [hunk 10]
```diff
@@ -1,12 +1,14 @@
...
 import net.corda.v5.base.annotations.CordaSerializable
 import net.corda.v5.ledger.utxo.StateAndRef
 
 @CordaSerializable
 data class UtxoLedgerTransactionContainer(
     val wireTransaction: WireTransaction,
     val inputStateAndRefs: List<StateAndRef<*>>,
-    val referenceStateAndRefs: List<StateAndRef<*>>
+    val referenceStateAndRefs: List<StateAndRef<*>>,
+    val groupParameters: SignedGroupParameters
```
Comment (vlajos) [inline libs/ledger/ledger-utxo-data/src/main/kotlin/net/corda/ledger/utxo/data/transaction/UtxoLedgerTransactionContainer.kt]: This class is used in serialized form in the Flow->Verification communication.
So it is possible that a 5.0 cluster sends a message, and then Corda gets stopped/upgraded/restarted
and then the Verification processor receives a UtxoLedgerTransactionContainer without `groupParameters`.
Could you check if this needs to be supported?
And then handle it accordingly.

File: components/virtual-node/sandbox-group-context-service/test.bndrun (hunk 8 above)
Comment (jennyang-r3) [reply components/virtual-node/sandbox-group-context-service/test.bndrun:12]: This resolves OSGi:resolve failure

File: components/ledger/ledger-verification/build.gradle (hunk 6 above)
Comment (jennyang-r3) [reply components/ledger/ledger-verification/build.gradle:48]: this module contains `GroupParametersFactory` that I used for the integration test and in `VerificationRequestHandlerImpl.kt` otherwise I get this compile error:
```
e: file:///Users/eunjee.yang/corda/corda-runtime-os/components/ledger/ledger-verification/src/main/kotlin/net/corda/ledger/verification/processor/impl/VerificationRequestHandlerImpl.kt:88:17 Cannot access class 'net.corda.membership.lib.SignedGroupParameters'. Check your module classpath for missing or conflicting dependencies
```

File: components/virtual-node/sandbox-group-context-service/test.bndrun (hunk 8 above)
Comment (jennyang-r3) [reply components/virtual-node/sandbox-group-context-service/test.bndrun:12]: I had this error before adding this:
```
1009 actionable tasks: 97 executed, 20 from cache, 892 up-to-date
Resolution failed. Capabilities satisfying the following requirements could not be found:
    [<<INITIAL>>]
      ⇒ osgi.identity: (osgi.identity=sandbox-group-context-service-tests)
          ⇒ [sandbox-group-context-service-tests version=5.1.0.0-SNAPSHOT]
              ⇒ osgi.wiring.package: (osgi.wiring.package=net.corda.ledger.utxo.data.state)
                  ⇒ [net.corda.ledger-utxo-data version=5.1.0.0-SNAPSHOT]
                      ⇒ osgi.service: (objectClass=net.corda.membership.lib.GroupParametersFactory)
                          ⇒ [net.corda.membership-impl version=5.1.0.0-SNAPSHOT]
                              ⇒ osgi.service: (objectClass=net.corda.avro.serialization.CordaAvroSerializationFactory)
The following requirements are optional:
    [net.corda.cipher-suite-impl version=5.1.0.0-SNAPSHOT]
      ⇒ osgi.service: (objectClass=net.corda.crypto.core.DigestAlgorithmFactoryProvider)
    [org.apache.commons.commons-compress version=1.22.0]
      ⇒ osgi.wiring.package: (osgi.wiring.package=com.github.luben.zstd)
      ⇒ osgi.wiring.package: (osgi.wiring.package=org.objectweb.asm)
      ⇒ osgi.wiring.package: (osgi.wiring.package=org.brotli.dec)
      ⇒ osgi.wiring.package: (osgi.wiring.package=org.tukaani.xz)
    [org.apache.felix.configadmin version=1.9.26]
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.osgi.service.log)(version>=1.3.0)(!(version>=2.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.osgi.service.coordinator)(version>=1.0.0)(!(version>=2.0.0)))
      ⇒ osgi.service: (objectClass=org.osgi.service.log.LogService)
    [net.corda.layered-property-map version=5.1.0.0-SNAPSHOT]
      ⇒ osgi.service: (objectClass=net.corda.layeredpropertymap.CustomPropertyConverter)
    [junit-platform-engine version=1.9.2]
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.apiguardian.api)(version>=1.1.0)(!(version>=2.0.0)))
    [junit-jupiter-engine version=5.9.2]
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.apiguardian.api)(version>=1.1.0)(!(version>=2.0.0)))
    [com.networknt.json-schema-validator version=1.0.79]
      ⇒ osgi.wiring.package: (osgi.wiring.package=org.joni)
      ⇒ osgi.wiring.package: (osgi.wiring.package=org.joni.exception)
      ⇒ osgi.wiring.package: (osgi.wiring.package=org.jcodings.specific)
      ⇒ osgi.wiring.package: (osgi.wiring.package=org.jcodings)
    [junit-platform-commons version=1.9.2]
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.apiguardian.api)(version>=1.1.0)(!(version>=2.0.0)))
    [junit-platform-launcher version=1.9.2]
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.apiguardian.api)(version>=1.1.0)(!(version>=2.0.0)))
    [junit-jupiter-api version=5.9.2]
      ⇒ osgi.wiring.package: (osgi.wiring.package=kotlin.jvm.internal)
      ⇒ osgi.wiring.package: (osgi.wiring.package=kotlin)
      ⇒ osgi.wiring.package: (osgi.wiring.package=kotlin.collections)
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.apiguardian.api)(version>=1.1.0)(!(version>=2.0.0)))
      ⇒ osgi.wiring.package: (osgi.wiring.package=kotlin.jvm.functions)
    [com.google.guava version=32.1.1.jre]
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=javax.annotation)(version>=3.0.0)(!(version>=4.0.0)))
    [net.corda.sandbox-amqp version=5.1.0.0-SNAPSHOT]
      ⇒ osgi.service: (objectClass=net.corda.serialization.InternalCustomSerializer)
    [net.corda.sandbox-group-context-service version=5.1.0.0-SNAPSHOT]
      ⇒ corda.system: (&(corda.system=*)(version>=1.0.0)(!(version>=2.0.0)))
    [net.corda.test-impl-one version=5.1.0.0-SNAPSHOT]
      ⇒ osgi.service: (objectClass=net.corda.v5.testing.NoSuchService)
    [junit-jupiter-params version=5.9.2]
      ⇒ osgi.wiring.package: (osgi.wiring.package=kotlin.jvm.internal)
      ⇒ osgi.wiring.package: (osgi.wiring.package=kotlin)
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.apiguardian.api)(version>=1.1.0)(!(version>=2.0.0)))
    [org.apache.felix.scr version=2.2.6]
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.osgi.service.metatype)(version>=1.2.0)(!(version>=2.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.osgi.service.log)(version>=1.4.0)(!(version>=2.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.apache.felix.service.command)(version>=1.0.0)(!(version>=2.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.osgi.service.cm)(version>=1.6.0)(!(version>=2.0.0)))
    [net.corda.test-utilities version=5.1.0.0-SNAPSHOT]
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=io.micrometer.core.instrument)(version>=1.11.0)(!(version>=2.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.junit.jupiter.api.extension)(version>=5.9.0)(!(version>=6.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.junit.jupiter.api)(version>=5.9.0)(!(version>=6.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=net.corda.lifecycle)(version>=5.1.0)(!(version>=6.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=io.micrometer.core.instrument.composite)(version>=1.11.0)(!(version>=2.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=net.corda.libs.packaging.core)(version>=5.1.0)(!(version>=6.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=net.corda.libs.cpi.datamodel)(version>=5.1.0)(!(version>=6.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=net.corda.v5.base.types)(version>=5.1.0)(!(version>=6.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.mockito.invocation)(version>=5.3.0)(!(version>=6.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=net.corda.cpiinfo.read)(version>=5.1.0)(!(version>=6.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=net.corda.crypto.core)(version>=5.1.0)(!(version>=6.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=net.corda.utilities.time)(version>=5.1.0)(!(version>=6.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=net.corda.metrics)(version>=5.1.0)(!(version>=6.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=net.corda.virtualnode)(version>=5.1.0)(!(version>=6.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=io.micrometer.core.instrument.simple)(version>=1.11.0)(!(version>=2.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=net.corda.v5.crypto)(version>=5.1.0)(!(version>=6.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.mockito.stubbing)(version>=5.3.0)(!(version>=6.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.mockito)(version>=5.3.0)(!(version>=6.0.0)))
    [bcprov version=1.73.0]
      ⇒ osgi.wiring.package: (osgi.wiring.package=junit.framework)
    [net.corda.test-impl-two version=5.1.0.0-SNAPSHOT]
      ⇒ osgi.service: (objectClass=net.corda.v5.testing.NoSuchService)
    [net.corda.sandbox-testing version=5.1.0.0-SNAPSHOT]
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.junit.jupiter.api.extension)(version>=5.9.0)(!(version>=6.0.0)))
    [assertj-core version=3.24.2]
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.opentest4j)(version>=1.2.0)(!(version>=2.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.junit.jupiter.api.extension)(version>=5.9.0)(!(version>=6.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.junit.jupiter.api)(version>=5.9.0)(!(version>=6.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.junit.platform.commons.support)(version>=1.9.0)(!(version>=2.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.junit.jupiter.api.parallel)(version>=5.9.0)(!(version>=6.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.hamcrest)(version>=2.2.0)(!(version>=3.0.0)))
      ⇒ osgi.wiring.package: (osgi.wiring.package=org.junit.rules)
      ⇒ osgi.wiring.package: (osgi.wiring.package=org.junit.runners.model)
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.junit.platform.commons.annotation)(version>=1.9.0)(!(version>=2.0.0)))
      ⇒ osgi.wiring.package: (osgi.wiring.package=org.junit.runner)
    [avro version=1.11.2]
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=com.github.luben.zstd)(version>=1.5.0)(!(version>=2.0.0)))
      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.xerial.snappy)(version>=1.1.0)(!(version>=2.0.0)))

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':components:virtual-node:sandbox-group-context-service:resolve'.
> /Users/eunjee.yang/corda/corda-runtime-os/components/virtual-node/sandbox-group-context-service/build/resolved-test.bndrun resolution exception

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.

* Get more help at https://help.gradle.org/

BUILD FAILED in 54s

Publishing build scan...
https://gradle.dev.r3.com/s/lic2zjjzo5n6q

17:23:09: Execution finished 'resolve'.
```

File: libs/ledger/ledger-utxo-data/build.gradle (hunk 7 above)
Comment (jennyang-r3) [reply libs/ledger/ledger-utxo-data/build.gradle]: It looks like we don't need this dependency.

File: components/virtual-node/sandbox-group-context-service/test.bndrun (hunk 8 above)
Comment (vlajos) [reply components/virtual-node/sandbox-group-context-service/test.bndrun:12]: Thanks.

File: components/ledger/ledger-verification/build.gradle (hunk 6 above)
Comment (vlajos) [reply components/ledger/ledger-verification/build.gradle:48]: Thanks.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/verifier/UtxoLedgerTransactionVerificationServiceImpl.kt (hunk 4 above)
Comment (jennyang-r3) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/verifier/UtxoLedgerTransactionVerificationServiceImpl.kt]: Does that mean I should inject `GroupParametersService` to `UtxoLedgerTransactionFactoryImpl` to get `currentGroupParameters`?

Comment (vlajos) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/verifier/UtxoLedgerTransactionVerificationServiceImpl.kt]: Yes. I think so.

File: libs/ledger/ledger-utxo-data/src/main/kotlin/net/corda/ledger/utxo/data/transaction/UtxoLedgerTransactionContainer.kt (hunk 10 above)
Comment (simon-johnson-r3) [reply libs/ledger/ledger-utxo-data/src/main/kotlin/net/corda/ledger/utxo/data/transaction/UtxoLedgerTransactionContainer.kt]: After some offline discussion, the usual rules of AMQP message evolution should be followed, which can be found here: https://docs.r3.com/en/platform/corda/5.0/developing-applications/api/amqp-serialization-default-evolution.html

In this case perhaps you would like to consider the choice of "adding nullable properties", otherwise please follow the alternative guidance on that page.

> And then handle it accordingly.

This is a domain specific issue so I'll leave you to handle it as you see fit based on some of the ideas we touched on.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/factory/UtxoLedgerTransactionFactory.kt [hunk 11]
```diff
@@ -27,4 +28,6 @@ interface UtxoLedgerTransactionFactory {
         inputStateAndRefs: List<UtxoTransactionOutputDto>,
         referenceStateAndRefs: List<UtxoTransactionOutputDto>
     ): UtxoLedgerTransactionInternal
+
+    fun getGroupParameters(wireTransaction: WireTransaction): GroupParameters
```
Comment (vlajos) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/factory/UtxoLedgerTransactionFactory.kt]: Does this need to be on this interface? Could not this be private?

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/verifier/UtxoLedgerTransactionVerificationServiceImpl.kt [hunk 12]
```diff
@@ -53,7 +46,7 @@ class UtxoLedgerTransactionVerificationServiceImpl @Activate constructor(
     override fun verify(transaction: UtxoLedgerTransaction) {
         recordSuspendable(::transactionVerificationFlowTimer) @Suspendable {
 
-            val signedGroupParameters = fetchAndVerifySignedGroupParameters(transaction)
+            val signedGroupParameters = verifySignedGroupParameters(transaction)
```
Comment (vlajos) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/verifier/UtxoLedgerTransactionVerificationServiceImpl.kt]: I think this function could be just inlined now.

File: libs/ledger/ledger-utxo-data/build.gradle [hunk 13]
```diff
@@ -14,6 +14,8 @@ dependencies {
 
     implementation project(':libs:crypto:crypto-core')
     implementation project(':libs:ledger:ledger-common-data')
+    implementation project(':libs:membership:membership-common')
+    implementation project(':libs:membership:membership-impl')
```
Comment (vlajos) [inline libs/ledger/ledger-utxo-data/build.gradle]: Is the `impl` needed as well?

Comment (jennyang-r3) [reply libs/ledger/ledger-utxo-data/build.gradle]: Without it, I get this resolution error. But not sure where exactly it is used I think it was implicit access from `GroupParametersFactory`
```
Resolution failed. Capabilities satisfying the following requirements could not be found:
    [<<INITIAL>>]
      ⇒ osgi.identity: (osgi.identity=ledger-persistence-tests)
          ⇒ [ledger-persistence-tests version=5.1.0.0-SNAPSHOT]
              ⇒ osgi.wiring.package: (osgi.wiring.package=net.corda.ledger.persistence.consensual.impl)
                  ⇒ [net.corda.ledger-persistence version=5.1.0.0-SNAPSHOT]
                      ⇒ osgi.wiring.package: (&(osgi.wiring.package=net.corda.ledger.utxo.data.transaction)(version>=5.1.0)(!(version>=6.0.0)))
                          ⇒ [net.corda.ledger-utxo-data version=5.1.0.0-SNAPSHOT]
                              ⇒ osgi.service: (objectClass=net.corda.membership.lib.GroupParametersFactory)
```

Comment (vlajos) [reply libs/ledger/ledger-utxo-data/build.gradle]: Is it needed only for integration tests?
Could not changing it to `integrationTestRuntimeOnly` be an option?

Comment (jennyang-r3) [reply libs/ledger/ledger-utxo-data/build.gradle]: as far as I know, I didn't touch any tests under this module. so I don't think it will be it

Comment (jennyang-r3) [reply libs/ledger/ledger-utxo-data/build.gradle]: I will have a look if there is a way to avoid adding this dependency to this module

Comment (lankydan) [reply libs/ledger/ledger-utxo-data/build.gradle]: I think what is going on here is.

- `SignedGroupParametersSerializer` was moved to `ledger-utxo-data`.
- @jennyang-r3 added the dependencies to support this move.

I think we can remove the `membership-impl` dependency from `ledger-utxo-data`. The `membership-common` should achieve what you need afaik.

You then need to add `integrationTestRuntimeOnly project(':libs:membership:membership-impl')` to `:components:ledger:ledger-persistence` to resolve the issue with the tests.

File: simulator/runtime/src/main/kotlin/net/corda/simulator/runtime/ledger/utxo/UtxoLedgerTransactionBase.kt (hunk 9 above)
Comment (jennyang-r3) [reply simulator/runtime/src/main/kotlin/net/corda/simulator/runtime/ledger/utxo/UtxoLedgerTransactionBase.kt:158]: https://r3holdco.slack.com/archives/C03KXQ89TC1/p1692697514364299

File: libs/ledger/ledger-utxo-data/src/main/kotlin/net/corda/ledger/utxo/data/transaction/UtxoLedgerTransactionImpl.kt [hunk 14]
```diff
@@ -115,6 +117,16 @@ class UtxoLedgerTransactionImpl(
         return outputContractStates.filterIsInstance(type)
     }
 
+    /*
+    * The null check in the method is to prevent 5.0 from hitting this method since this field doesn't exist in 5.0
```
Comment (vlajos) [inline libs/ledger/ledger-utxo-data/src/main/kotlin/net/corda/ledger/utxo/data/transaction/UtxoLedgerTransactionImpl.kt]:
```suggestion
    * The null check in the method is to make 5.0 in-flight transactions caught in upgrades failing more gracefully.
```

If this is correct at all.

**Review by dickon — DISMISSED** [review]
LGTM for sandbox changes

**Review by charlieR3 — DISMISSED** [review]
LGTM - from a network team perspective

**Review by ronanbrowne — DISMISSED** [review]
LGTM - build files. but you still need to change

cordaApiVersion=5.1.0.16-alpha-1692879578824

**Review by vlajos — DISMISSED** [review]

**Review by charlieR3 — DISMISSED** [review]

**Review by pmck91 — DISMISSED** [review]
LGTM flow perspective

**Review by vlajos — APPROVED** [review]

**Review by charlieR3 — APPROVED** [review]

**Review by lankydan — APPROVED** [review]

**Review by pmck91 — APPROVED** [review]

**Review by ronanbrowne — APPROVED** [review]

**Review by vlajos — APPROVED** [review]

**Review by dickon — APPROVED** [review]
LGTM
